apollo-router/src/query_planner/bridge_query_planner.rs

```diff
@@ -191,45 +191,37 @@ impl BridgeQueryPlanner {
-                            match diff_line {
-                                diff::Result::Left(l) => {
-                                    let trimmed = l.trim();
-                                    if !trimmed.starts_with('#') && !trimmed.is_empty() {
-                                        writeln!(&mut output, "-{l}")
-                                            .expect("write will never fail");
-                                    } else {
-                                        writeln!(&mut output, " {l}")
-                                            .expect("write will never fail");
-                                    }
-                                }
-                                diff::Result::Both(l, _) => {
-                                    writeln!(&mut output, " {l}").expect("write will never fail");
-                                }
-                                diff::Result::Right(r) => {
-                                    let trimmed = r.trim();
-                                    if trimmed != "---" && !trimmed.is_empty() {
-                                        writeln!(&mut output, "+{r}")
-                                            .expect("write will never fail");
-                                    }
-                                }
+                    (Ok(left), Ok(right)) => {
+                        // To compare results, we re-parse, standardize, and print with apollo-rs,
+                        // so the formatting is identical.
+                        if let (Ok(left), Ok(right)) = (
+                            apollo_compiler::Schema::parse(left, "js.graphql"),
+                            apollo_compiler::Schema::parse(right, "rust.graphql"),
+                        ) {
+                            let left = standardize_schema(left).to_string();
+                            let right = standardize_schema(right).to_string();
+
+                            if left == right {
+                                tracing::warn!(
+                                    monotonic_counter.apollo.router.lifecycle.api_schema = 1u64,
+                                    generation.is_matched = true,
+                                );
+                            } else {
+                                tracing::warn!(
+                                    monotonic_counter.apollo.router.lifecycle.api_schema = 1u64,
+                                    generation.is_matched = false,
+                                    "API schema generation mismatch: apollo-federation and router-bridge write different schema"
+                                );
+
+                                let differences = diff::lines(&left, &right);
+                                tracing::debug!(
+                                    "different API schema between apollo-federation and router-bridge:\n{}",
+                                    render_diff(&differences),
+                                );
```

Comment:
how do we get to see the diff for customer schemas? if we spot a mismatch in our metrics, should we reach out to people to give us this debug log? i don't mind this as it does get us talking to people who could be potentially helpful when we start swapping out the query planner, just want to make sure we are on the same page on how we will get to this log.

Reply:
Yeah we won't see this log. But since this is about supergraph schemas, we can very easily collect them from Studio and run the comparison ourselves